### PR TITLE
Deviantart Ripper changes

### PR DESCRIPTION
**Pull request by MrPlaygon**
# Category

This change is exactly one of the following (please change `[ ]` to `[x]`) to indicate which:
* [X] a bug fix (Fix #...)
* [ ] a new Ripper
* [X] a refactoring
* [ ] a style change/fix
* [ ] a new feature

# Description

New Deviantart Ripper. HTML based.
NOT using Deviantart API like the old JSON ripper because it is SLOW (and IMHO annoying to use)

Things to consider: Using the API might be less work/maintenance later because APIs do not change as frequently as HTML source code...?

# Testing

Required verification:
* [ ] I've verified that there are no regressions in `mvn test` (there are no new failures or errors).
 -> had to uncomment something because of new design
* [ ] I've verified that this change works as intended.
  * [X] Downloads all relevant content.
  * [X] Downloads content from multiple pages (as necessary or appropriate).
  * [X] Saves content at reasonable file names (e.g. page titles or content IDs) to help easily browse downloaded content.
* [X] I've verified that this change did not break existing functionality (especially in the Ripper I modified).
 -> can basically download galleries, favorites, specific (gallery/fav)folders. Tested with JPG, PNG and PDF
Optional but recommended:
* [ ] I've added a unit test to cover my change.

### CODE REVIEW

**Comment by MrPlaygon** [comment]
Yeah that is a very good idea. Will include that tomorrow!
Still unsure about not using the API. What is your opinion on that?

**Comment by rautamiekka** [comment]
> Yeah that is a very good idea. Will include that tomorrow!
> Still unsure about not using the API. What is your opinion on that?

The API is a real fucking disaster. Not using it is the right course for the time being.

Someone here did mention they offer the OAuth2 API in addition to their own one we're talking about. There's not much of DA-provided documentation on it, and the whole official API documentation just assumes you know everything already, but at least it seems a little bit more viable by the looks of it.

Then again, why bother with possibly useless API when you need to do HTML scraping to gain some of the info anyway ?

File: src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java [hunk 1]
```diff
@@ -1,409 +1,531 @@
...
+			// Load login page
+			Response res = Http.url("https://www.deviantart.com/users/login").connection().method(Method.GET)
+					.referrer(referer).userAgent(userAgent).execute();
+
+			// Find tokens
+			Document doc = res.parse();
+			Element form = doc.getElementById("login");
+			String token = form.select("input[name=\"validate_token\"]").first().attr("value");
+			String key = form.select("input[name=\"validate_key\"]").first().attr("value");
+			System.out.println(
```
Comment (cyian-1756) [inline src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java]: Please use LOGGER for logging and not System.out.println(). LOGGER logs both to the cli and the log file.

Also please avoid using dashes for formatting

File: src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java [hunk 2]
```diff
@@ -1,409 +1,531 @@
...
+			res = Http.url(res.header("location")).connection().referrer(referer).userAgent(userAgent)
+					.method(Method.GET).cookies(cookies).followRedirects(false).execute();
+
+			// Store cookies
+			updateCookie(res.cookies());
+
+			// Apply agegate
+			this.cookies.put("agegate_state", "1");
+
+			// Write Cookie to file for other RipMe Instances
```
Comment (cyian-1756) [inline src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java]: Why write a file to disk here? You could just set a value in ripmes config file. There's Utils.setConfigString("DeviantartRipper.cookies", cookies.toString()) for setting it and AFAIK there is another util func for turning cookie strings back into cookies

**Comment by MrPlaygon** [comment]
I will look into this. Thanks :)
Might take some time because I'm on vacation.

**Comment by rautamiekka** [comment]
Speaking of utils, please allow loading the logins from the config.

And when you're doing that, ensure to catch when the used username has been blocked by the gallery owner so you can
1) log/report it
2) either ...
2.1) switch over to the default one, if using custom logins
2.2) stop ripping that gallery and switch to the next, if a next one has been provided

If it's even possible, each gallery link provided to RipMe would optimally be checked right after logging in to see if the logins are blocked or if the owner hasn't been deactivated. I can help with that by providing the relevant HTML code.

**Comment by MrPlaygon** [comment]
Sounds good. Is there some form of documentation regarding utils and what can be done with it?

Furthermore I played with storing cookies as a string and I think I got it working but I have not figured out how to check if the cookies are expired. Will solve this when back at home :)

**Comment by rautamiekka** [comment]
As I keep reading the code, IIUCTI, it seems you're initiating a new connection whenever you're fetching a page. Right ?

If that's the case, you should look into reusing existing connections to the same target, but making a new connection if the target is somewhere else, such as those Wix CDN. That way DA has much less incentive to auto-ban, since the expensive price of initiating a new HTTP connection is avoided, which also makes life for the Client easier.

The scanning of a gallery/scraps/favs is a particularly good area where reusing a connection makes perfect sense.

File: src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java (hunk 1 above)
Comment (MrPlaygon) [reply src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java]: Using LOGGER now.

File: src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java (hunk 2 above)
Comment (MrPlaygon) [reply src/main/java/com/rarchives/ripme/ripper/rippers/DeviantartRipper.java]: Login Cookie is now stores as BASE64 encoded String in config file.

**Comment by MrPlaygon** [comment]
> 
> 
> Speaking of utils, please allow loading the logins from the config.
> 
> And when you're doing that, ensure to catch when the used username has been blocked by the gallery owner so you can
> 
>     1. log/report it
> 
>     2. either ...
>        2.1) switch over to the default one, if using custom logins
>        2.2) stop ripping that gallery and switch to the next, if a next one has been provided
> 
> 
> If it's even possible, each gallery link provided to RipMe would optimally be checked right after logging in to see if the logins are blocked or if the owner hasn't been deactivated, before ripping. The steps above apply to this as well, except for 2.2 that since you're unable to rip using both a custom and the default there's nothing next, apart from proceeding to the other links the ripper doesn't care about. I can help with that by providing the relevant HTML code.

This will take a LOT of time to test and figure out because I need to create deviantart accounts just to test how the website behaves when a user is blocked. Same for deactivated profiles or similar rare stuff.
I will see what I can do.

**Comment by rautamiekka** [comment]
> I will see what I can do.
When I did my research for my own software, I found out these:

For deactivated and banned accs (there's no way to distinguish between a self-deac and a banned), the HTML contains `This account is inactive.`. For example, https://www.deviantart.com/ghostofflossenburg is banned and https://www.deviantart.com/gingerbreadpony is self-deac.

For ones blocked by either party, the HTML contains `Sorry, you are unable to view this profile.`.

Those are the things I figured are the most reliable and static things to see deac/ban/block of an acc. There's a `<meta property="og:title" content="... | DeviantArt">` tag for each case, but I consider them unreliable.

**Comment by MrPlaygon** [comment]
> For ones blocked by either party, the HTML contains `Sorry, you are unable to view this profile.`. I suppose you could look for all-lowercase `you are unable to view this profile.` instead, so a subtle change in the text shouldn't require correction.

Deviantart returns a status code 404. Might use this one.

**Comment by rautamiekka** [comment]
> > For ones blocked by either party, the HTML contains `Sorry, you are unable to view this profile.`. I suppose you could look for all-lowercase `you are unable to view this profile.` instead, so a subtle change in the text shouldn't require correction.
> 
> Deviantart returns a status code 404. Might use this one.

Oh right, forgot that one, but when I found out about that I figured it's not exactly very reliable either.

**Comment by cyian-1756** [comment]
Looks good to me. I'll merge this once I get a chance to test it

**Comment by cyian-1756** [comment]
Looking at the tests it seems that the ripper now fails on some urls (https://www.deviantart.com/faterkcx/gallery/). Once that is fixed I'll merge this in!

**Comment by MrPlaygon** [comment]
Will look into it.

**Comment by MrPlaygon** [comment]
I saw an error, too but cannot identify why. The one I got was a 404 HTTP Error when trying to download an image. However when I copy the URL in my browser it works. Confused right now...
Do you have any ideas?

**Comment by rautamiekka** [comment]
Oh I fucking hate GitHub's ambiguousness ... I just wanted to comment without participating in the review I started earlier, so I tried to back off from what seemed like that, but instead I totally erased that review I started before -.-

> I saw an error, too but cannot identify why. The one I got was a 404 HTTP Error when trying to download an image. However when I copy the URL in my browser it works. Confused right now...
> Do you have any ideas?

Are the links downloaded immediately or postponed ? DA doesn't like postponing.

**Comment by cyian-1756** [comment]
> I saw an error, too but cannot identify why. The one I got was a 404 HTTP Error when trying to download an image. However when I copy the URL in my browser it works. Confused right now...
Do you have any ideas?

It might be cookies. Open your browsers dev tools, goto the network tab, load the page and then inspect the request

**Comment by MrPlaygon** [comment]
> Oh I fucking hate GitHub's ambiguousness ... I just wanted to comment without participating in the review I started earlier, so I tried to back off from what seemed like that, but instead I totally erased that review I started before -.-
> 
> > I saw an error, too but cannot identify why. The one I got was a 404 HTTP Error when trying to download an image. However when I copy the URL in my browser it works. Confused right now...
> > Do you have any ideas?
> 
> Are the links downloaded immediately or postponed ? DA doesn't like postponing.

Links are postponed but not long. I do not open them too late because after the error the link works fine. Maybe I am trying to open the link too fast? :o

**Comment by MrPlaygon** [comment]
> > I saw an error, too but cannot identify why. The one I got was a 404 HTTP Error when trying to download an image. However when I copy the URL in my browser it works. Confused right now...
> > Do you have any ideas?
> 
> It might be cookies. Open your browsers dev tools, goto the network tab, load the page and then inspect the request

Cookies are note the problem. No cookies required for the final download URL.

**Comment by rautamiekka** [comment]
> > Are the links downloaded immediately or postponed ? DA doesn't like postponing.
> 
> Links are postponed but not long. I do not open them too late because after the error the link works fine. Maybe I am trying to open the link too fast? :o

_Ugh, DA's quirks_ ... Maybe.

**Comment by MrPlaygon** [comment]
> > > Are the links downloaded immediately or postponed ? DA doesn't like postponing.
> > 
> > 
> > Links are postponed but not long. I do not open them too late because after the error the link works fine. Maybe I am trying to open the link too fast? :o
> 
> _Ugh, DA's quirks_ ... Maybe.

Nevertheless I added the DA Cookies to the final download as well.
Will try to implement a retry to see if it is really a timing thing or another problem...

**Comment by cyian-1756** [comment]
> Cookies are note the problem

What about the referrer?

**Comment by MrPlaygon** [comment]
Hey guys,
I tried the mentioned gallery a few more times and I had no error. I got almost 2000 images. Could you try again, please?

**Comment by rautamiekka** [comment]
> Hey guys,
> I tried the mentioned gallery a few more times and I had no error. I got almost 2000 images. Could you try again, please?

I can't test without a new release cuz I can't figure out Eclipse. I tried with 1.7.81 and got 490 using https://www.deviantart.com/faterkcx/gallery/?catpath=/

**Comment by cyian-1756** [comment]
Running this I get the following error 

```
java.lang.NullPointerException
	at com.rarchives.ripme.ripper.rippers.DeviantartRipper.updateCookie(DeviantartRipper.java:402)
	at com.rarchives.ripme.ripper.rippers.DeviantartRipper.login(DeviantartRipper.java:164)
	at com.rarchives.ripme.ripper.rippers.DeviantartRipper.getFirstPage(DeviantartRipper.java:127)
	at com.rarchives.ripme.ripper.AbstractHTMLRipper.rip(AbstractHTMLRipper.java:86)
	at com.rarchives.ripme.ripper.AbstractRipper.run(AbstractRipper.java:562)
	at java.lang.Thread.run(Thread.java:748)
```

I tried ripping https://www.deviantart.com/faterkcx/gallery/

It should be simple neough to fix by adding a 

```
if (m == null) {return;}
```

to line 401

**Comment by MrPlaygon** [comment]
Aaah I see now. Yes you are right. My bad.

**Comment by cyian-1756** [comment]
There's some minor unittests failing, but nothing blocking so I'll merge this

**Comment by rautamiekka** [comment]
> There's some minor unittests failing, but nothing blocking so I'll merge this

This might be glorious ... :D

**Pull request merged**